plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarConnector.java

```diff
@@ -0,0 +1,96 @@
+ * limitations under the License.
+ */
+package io.trino.plugin.pulsar;
+
+import io.airlift.bootstrap.LifeCycleManager;
+import io.airlift.bootstrap.LifeCycleStopException;
+import io.airlift.log.Logger;
+import io.trino.plugin.base.classloader.ClassLoaderSafeConnectorMetadata;
+import io.trino.spi.connector.Connector;
+import io.trino.spi.connector.ConnectorMetadata;
+import io.trino.spi.connector.ConnectorRecordSetProvider;
+import io.trino.spi.connector.ConnectorSplitManager;
+import io.trino.spi.connector.ConnectorTransactionHandle;
+import io.trino.spi.transaction.IsolationLevel;
+
+import javax.inject.Inject;
+
+import static io.trino.spi.transaction.IsolationLevel.READ_COMMITTED;
+import static io.trino.spi.transaction.IsolationLevel.checkConnectorSupports;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Implementation of Pulsar connector for Trino.
+ */
+public class PulsarConnector
+        implements Connector
+{
+    private final LifeCycleManager lifeCycleManager;
+    private final PulsarMetadata metadata;
+    private final PulsarSplitManager splitManager;
+    private final PulsarRecordSetProvider recordSetProvider;
+    private final PulsarConnectorConfig pulsarConnectorConfig;
+
+    @Inject
+    public PulsarConnector(
+            LifeCycleManager lifeCycleManager,
+            PulsarMetadata metadata,
+            PulsarSplitManager splitManager,
+            PulsarRecordSetProvider recordSetProvider,
+            PulsarConnectorConfig pulsarConnectorConfig)
+    {
+        this.lifeCycleManager = requireNonNull(lifeCycleManager, "lifeCycleManager is null");
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.splitManager = requireNonNull(splitManager, "splitManager is null");
+        this.recordSetProvider = requireNonNull(recordSetProvider, "recordSetProvider is null");
+        this.pulsarConnectorConfig = requireNonNull(pulsarConnectorConfig, "pulsarConnectorConfig is null");
+    }
+
+    @Override
+    public ConnectorTransactionHandle beginTransaction(IsolationLevel isolationLevel, boolean readOnly)
+    {
+        checkConnectorSupports(READ_COMMITTED, isolationLevel);
+        return PulsarTransactionHandle.INSTANCE;
+    }
+
+    @Override
+    public ConnectorMetadata getMetadata(ConnectorTransactionHandle transactionHandle)
+    {
+        return new ClassLoaderSafeConnectorMetadata(metadata, getClass().getClassLoader());
+    }
+
+    @Override
+    public ConnectorSplitManager getSplitManager()
+    {
+        return splitManager;
+    }
+
+    @Override
+    public ConnectorRecordSetProvider getRecordSetProvider()
+    {
+        return recordSetProvider;
+    }
+
+    public void initConnectorCache() throws Exception
+    {
+        PulsarConnectorCacheImpl.getConnectorCache(pulsarConnectorConfig);
+    }
+
+    @Override
+    public final void shutdown()
+    {
+        pulsarConnectorConfig.close();
```

Comment:
Config class should be plain beans, they should not be closeable. I think you wanted to have a separate entity that could be closed. 

Also you don't have to do it here. You can use `@PostConstruct` annotation.

Reply:
It's a Admin client talking to Pulsar used to get metadata for topic, we only want to close it when we're done using the connector, so removed it to a separate class but still keep its `close` method in `shutdown`.